File: GhostChat-iOS/GhostChat/Features/Contacts/ContactsViewModel.swift
import Foundation

/// ViewModel for contacts list — fetches from ContactStore (SQLCipher)
@MainActor
final class ContactsViewModel: ObservableObject {

    @Published var contacts: [Contact] = []
    @Published var errorMessage: String?
    @Published var lastMessages: [String: ChatMessage] = [:]   // contactId -> last message
    @Published var unreadCounts: [String: Int] = [:]           // contactId -> unread count

    private let store = ContactStore()
    private let messageStore = MessageStore()

    func loadContacts() {
        do {
            contacts = try store.fetchAll()
            loadMessagePreviews()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func loadMessagePreviews() {
        var previews: [String: ChatMessage] = [:]
        var counts: [String: Int] = [:]
        for contact in contacts {
            let cId = contact.id.uuidString
            if let last = try? messageStore.fetchLastMessage(for: cId) {
                previews[cId] = last
            }
            if let count = try? messageStore.countUnread(for: cId), count > 0 {
                counts[cId] = count
            }
        }
        // Also load Saved Messages preview
        let savedId = ChatViewModel.savedMessagesContactId
        if let last = try? messageStore.fetchLastMessage(for: savedId) {
            previews[savedId] = last
        }
        lastMessages = previews
        unreadCounts = counts
    }

    func deleteContact(_ contact: Contact) {
        do {
            try store.delete(id: contact.id)
            try? messageStore.deleteForContact(contact.id.uuidString)
            contacts.removeAll { $0.id == contact.id }
            lastMessages.removeValue(forKey: contact.id.uuidString)
            unreadCounts.removeValue(forKey: contact.id.uuidString)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateContactLabel(_ contact: Contact, newLabel: String) {
        do {
            try store.updateLabel(contactId: contact.id, label: newLabel)
            if let idx = contacts.firstIndex(where: { $0.id == contact.id }) {
                contacts[idx].label = newLabel
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateContactNotes(_ contact: Contact, notes: String) {
        do {
            let trimmed = notes.trimmingCharacters(in: .whitespacesAndNewlines)
            try store.updateNotes(contactId: contact.id, notes: trimmed.isEmpty ? nil : trimmed)
            if let idx = contacts.firstIndex(where: { $0.id == contact.id }) {
                contacts[idx].notes = trimmed.isEmpty ? nil : trimmed
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateMessageTTL(_ contact: Contact, ttl: Int?) {
        do {
            try store.updateMessageTTL(contactId: contact.id, ttl: ttl)
            if let idx = contacts.firstIndex(where: { $0.id == contact.id }) {
                contacts[idx].messageTTL = ttl
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteAllContacts() {
        do {
            try messageStore.deleteAll()
            try store.deleteAll()
            contacts.removeAll()
            lastMessages.removeAll()
            unreadCounts.removeAll()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteHistory(for contact: Contact) {
        try? messageStore.deleteForContact(contact.id.uuidString)
        lastMessages.removeValue(forKey: contact.id.uuidString)
        unreadCounts.removeValue(forKey: contact.id.uuidString)
    }
}

File: GhostChat-Android/app/src/main/java/com/ghost/chat/features/contacts/ContactsViewModel.kt
package com.ghost.chat.features.contacts

import android.content.Context
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ghost.chat.core.storage.ContactStore
import com.ghost.chat.core.storage.DatabaseService
import com.ghost.chat.core.storage.MessageStore
import com.ghost.chat.models.ChatMessage
import com.ghost.chat.models.Contact
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ContactsViewModel(context: Context) : ViewModel() {

    private val db = DatabaseService.getInstance(context)
    private val contactStore = ContactStore(db)
    private val messageStore = MessageStore(db)
    val contacts = mutableStateListOf<Contact>()
    val lastMessages = mutableStateMapOf<String, ChatMessage>()   // contactId -> last message
    val unreadCounts = mutableStateMapOf<String, Int>()           // contactId -> unread count

    fun loadContacts() {
        viewModelScope.launch(Dispatchers.IO) {
            val loaded = contactStore.fetchAll()
            val previews = mutableMapOf<String, ChatMessage>()
            val counts = mutableMapOf<String, Int>()
            for (contact in loaded) {
                val cId = contact.id
                messageStore.fetchLastMessage(cId)?.let { previews[cId] = it }
                val count = messageStore.countUnread(cId)
                if (count > 0) counts[cId] = count
            }
            // Also load Saved Messages preview
            val savedId = com.ghost.chat.features.chat.ChatViewModel.SAVED_MESSAGES_CONTACT_ID
            messageStore.fetchLastMessage(savedId)?.let { previews[savedId] = it }
            withContext(Dispatchers.Main) {
                contacts.clear()
                contacts.addAll(loaded)
                lastMessages.clear()
                lastMessages.putAll(previews)
                unreadCounts.clear()
                unreadCounts.putAll(counts)
            }
        }
    }

    fun deleteContact(contact: Contact) {
        viewModelScope.launch(Dispatchers.IO) {
            messageStore.deleteForContact(contact.id)
            contactStore.delete(contact.id)
            withContext(Dispatchers.Main) {
                contacts.remove(contact)
                lastMessages.remove(contact.id)
                unreadCounts.remove(contact.id)
            }
        }
    }

    fun updateLabel(contact: Contact, newLabel: String) {
        viewModelScope.launch(Dispatchers.IO) {
            contactStore.updateLabel(contact.id, newLabel)
            withContext(Dispatchers.Main) {
                val index = contacts.indexOfFirst { it.id == contact.id }
                if (index >= 0) {
                    contacts[index] = contacts[index].copy(label = newLabel)
                }
            }
        }
    }

    fun updateNotes(contact: Contact, notes: String?) {
        viewModelScope.launch(Dispatchers.IO) {
            contactStore.updateNotes(contact.id, notes)
            withContext(Dispatchers.Main) {
                val index = contacts.indexOfFirst { it.id == contact.id }
                if (index >= 0) {
                    contacts[index] = contacts[index].copy(notes = if (notes.isNullOrBlank()) null else notes)
                }
            }
        }
    }

    fun updateMessageTTL(contact: Contact, ttl: Int?) {
        viewModelScope.launch(Dispatchers.IO) {
            contactStore.updateMessageTTL(contact.id, ttl)
            withContext(Dispatchers.Main) {
                val index = contacts.indexOfFirst { it.id == contact.id }
                if (index >= 0) {
                    contacts[index] = contacts[index].copy(messageTTL = ttl)
                }
            }
        }
    }

    fun deleteHistory(contact: Contact) {
        viewModelScope.launch(Dispatchers.IO) {
            messageStore.deleteForContact(contact.id)
            withContext(Dispatchers.Main) {
                lastMessages.remove(contact.id)
                unreadCounts.remove(contact.id)
            }
        }
    }
}
